### PR TITLE
Import

### PR DESCRIPTION
**Pull request by joel**

### CODE REVIEW

File: lib/dir_model/import.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
+module DirModel
+  module Import
+    extend ActiveSupport::Concern
+
+    attr_reader :context, :source_path, :index, :previous
+
+    def initialize(path, options={})
+      @source_path, @context = path, OpenStruct.new(options[:context])
+      @index, @previous = options[:index], options[:previous].try(:dup)
+      @load_state = :ghost
```
Comment (s12chung) [inline lib/dir_model/import.rb:10]: what is this for...?

Comment (s12chung) [inline lib/dir_model/import.rb:10]: never mind. i got it.

**Comment by s12chung** [comment]
focusing on the high level... the first thing i'd do is remove features---some features are `csv_row_model` aren't need. let's remove them 1 by 1 so we can revert later.

File: lib/dir_model/import.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
+module DirModel
+  module Import
+    extend ActiveSupport::Concern
+
+    attr_reader :context, :source_path, :index, :previous
+
+    def initialize(path, options={})
+      @source_path, @context = path, OpenStruct.new(options[:context])
+      @index, @previous = options[:index], options[:previous].try(:dup)
```
Comment (s12chung) [inline lib/dir_model/import.rb:9]: let's remove `previous`.

File: lib/dir_model/import/path.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        @next_path ||= _read_path
+      end
+
+      def read_path
+        @previous_path = @current_path
+        unless caching_value?
+          @current_path = _read_path { forward! }
+        end
+        current_path
+      end
```
Comment (s12chung) [inline lib/dir_model/import/path.rb:42]: not needed?

File: lib/dir_model/import/path.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        else
+          false
+        end
+      end
+
+      def ruby_path
+        @ruby_path ||= Pathname.glob("spec/#{path}/**/*").each
+      end
+
+      def _read_path
```
Comment (s12chung) [inline lib/dir_model/import/path.rb:61]: removing above will allow you to move this into `def next_path`

File: lib/dir_model/import/path.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          @next_path = nil
+          forward!
+          true
+        else
+          false
+        end
+      end
+
+      def ruby_path
+        @ruby_path ||= Pathname.glob("spec/#{path}/**/*").each
```
Comment (s12chung) [inline lib/dir_model/import/path.rb]: :warning:

pass the full path in the spec?

File: lib/dir_model/import/path.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      def caching_value?
+        if @next_path
+          @current_path = @next_path
+          @next_path = nil
+          forward!
+          true
+        else
+          false
+        end
+      end
```
Comment (s12chung) [inline lib/dir_model/import/path.rb:55]: removing `read_path` will remove this :)

File: lib/dir_model/import/path.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        yield if block_given?
+        path
+      rescue StopIteration
+        set_end
+       nil
+      end
+
+      def forward!
+        @index += 1
+      end
```
Comment (s12chung) [inline lib/dir_model/import/path.rb:72]: this won't be needed either, just in `def next_path`

File: lib/dir_model/import/path.rb (hunk 3 above)
Comment (s12chung) [inline lib/dir_model/import/path.rb:42]: i wish my `csv_row_model`'s `csv` code was simpler, but it had constraints that pushed it to be complicated. in this case, i needed this to say "if the next row is not a child, stop parsing children". you have not need to do a "look ahead" operation.

File: lib/dir_model/import/dir.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+      extend Forwardable
+
+      attr_reader :index
+      attr_reader :path
+      attr_reader :import_dir_model_class
+      attr_reader :context
+      attr_reader :current_dir_model
+      attr_reader :previous_dir_model
+
+      def_delegators :@path, :end?
```
Comment (s12chung) [inline lib/dir_model/import/dir.rb:15]: didn't know this existed until now.

File: lib/dir_model/export.rb [hunk 9]
```diff
@@ -25,18 +27,37 @@ def generated?
...
       @generated = true
     end
 
+    def get_value_of(string_or_proc)
+      return string_or_proc if string_or_proc.is_a?(String)
+      instance_exec(&string_or_proc)
+    end
+
+    def get_extension(file, extensions)
+      extension = FastImage.type(self.public_send(file))
```
Comment (s12chung) [inline lib/dir_model/export.rb]: hmm.... let's simplify and just use the `filename`? it's not always images (could be a json ;)

File: lib/dir_model/export.rb [hunk 10]
```diff
@@ -25,18 +27,37 @@ def generated?
...
+      if extensions.include?(extension) or
```
Comment (s12chung) [inline lib/dir_model/export.rb]: hmm... i didn't think of this before :+1:

File: spec/dir_model/import/dir_spec.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    end
+  end
+
+  describe '#reset' do
+    subject { instance.reset }
+
+    context 'at the end of paths' do
+      before { while instance.next; end }
+
+      it 'resets and starts at the first path' do
```
Comment (s12chung) [inline spec/dir_model/import/dir_spec.rb:27]: `resets`? why "and starts at the first path"?

**Comment by s12chung** [comment]
tests look good overall :white_check_mark:

File: lib/dir_model/export.rb (hunk 9 above)
Comment (joel) [inline lib/dir_model/export.rb]: In the export context we don't have any idea of which kind of document is provided by the source, we have a path of the destination and a file from the source `self.public_send(file_name).read` so here we expect client know which kind of file is provided here, so we need to ask explicitly the client about that.

File: lib/dir_model/export.rb (hunk 10 above)
Comment (joel) [inline lib/dir_model/export.rb]: I think is the wrong place this should be on import only

File: lib/dir_model/import/dir.rb (hunk 8 above)
Comment (joel) [inline lib/dir_model/import/dir.rb:15]: Playing without `ActiveSupport` :grin:

File: lib/dir_model/import/path.rb (hunk 5 above)
Comment (joel) [inline lib/dir_model/import/path.rb]: outch :flushed:

File: lib/dir_model/import/path.rb (hunk 3 above)
Comment (joel) [inline lib/dir_model/import/path.rb:42]: Yeah I can simplify this file indeed, but it was a lab to understand `CsvRowModel::Import::Csv` and `CsvRowModel::Import::File` but I'm pretty happy about the result, we have `previous`, `current` and `next` path, and I think is easier to understand than `csv_row_model` code IMO

File: lib/dir_model/export.rb (hunk 9 above)
Comment (joel) [inline lib/dir_model/export.rb]: Ok I remove all extension related stuff, just doesn't make sens at all....

**Pull request merged**